File: server/models/CreditScore.js
const mongoose = require("mongoose");

const creditScoreSchema = new mongoose.Schema(
  {
    msmeId: { type: mongoose.Schema.Types.ObjectId, ref: "MSME", required: true },
    generatedBy: { type: mongoose.Schema.Types.ObjectId, ref: "User", required: true },
    scoreValue: { type: Number, required: true, min: 300, max: 850 },
    riskCategory: {
      type: String,
      enum: ["Low", "Medium", "High"],
      required: true,
    },
    modelVersion: { type: String, default: "xgboost_v1" },
    shapValues: {
      gstConsistency: Number,
      cashFlowHealth: Number,
      revenueGrowth: Number,
      paymentBehaviour: Number,
      transactionVolume: Number,
      chequeBouncerate: Number,
    },
    shapSummary: [
      {
        feature: String,
        impact: Number,
        direction: { type: String, enum: ["positive", "negative"] },
        displayLabel: String,
      },
    ],
    featureInputSnapshot: { type: mongoose.Schema.Types.Mixed },
    recommendedLoanAmount: { type: Number },
    recommendedInterestBand: { type: String },
    eligibleGovernmentSchemes: [String],
    stressSignals: [
      {
        signal: String,
        severity: { type: String, enum: ["warning", "critical"] },
        description: String,
      },
    ],
    fraudFlags: [
      {
        flag: String,
        severity: { type: String, enum: ["medium", "high"] },
        description: String,
      },
    ],
    explanationText: { type: String },
    auditHash: { type: String },
  },
  { timestamps: true }
);

creditScoreSchema.index({ msmeId: 1, createdAt: -1 });
creditScoreSchema.index({ generatedBy: 1 });

module.exports = mongoose.model("CreditScore", creditScoreSchema);

File: server/app/models/credit_score.py
"""CreditScore document — mirrors the original Mongoose CreditScore schema."""

from beanie import Document
from pydantic import Field
from typing import Optional, Literal, Any
from datetime import datetime


class SHAPSummaryItem(dict):
    pass


class StressSignalItem(dict):
    pass


class FraudFlagItem(dict):
    pass


class CreditScore(Document):
    msme_id: str
    generated_by: str
    score_value: int = Field(ge=300, le=850)
    risk_category: Literal["Low", "Medium", "High"]
    model_version: str = "xgboost_v1"
    shap_values: Optional[dict] = None
    shap_summary: Optional[list] = None
    feature_input_snapshot: Optional[dict] = None
    recommended_loan_amount: Optional[float] = None
    recommended_interest_band: Optional[str] = None
    eligible_government_schemes: list[str] = []
    stress_signals: list[dict] = []
    fraud_flags: list[dict] = []
    explanation_text: Optional[str] = None
    audit_hash: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "creditscores"

    def to_response_dict(self) -> dict:
        return {
            "_id": str(self.id),
            "msmeId": self.msme_id,
            "generatedBy": self.generated_by,
            "scoreValue": self.score_value,
            "riskCategory": self.risk_category,
            "modelVersion": self.model_version,
            "shapValues": self.shap_values,
            "shapSummary": self.shap_summary,
            "featureInputSnapshot": self.feature_input_snapshot,
            "recommendedLoanAmount": self.recommended_loan_amount,
            "recommendedInterestBand": self.recommended_interest_band,
            "eligibleGovernmentSchemes": self.eligible_government_schemes,
            "stressSignals": self.stress_signals,
            "fraudFlags": self.fraud_flags,
            "explanationText": self.explanation_text,
            "auditHash": self.audit_hash,
            "createdAt": self.created_at.isoformat(),
            "updatedAt": self.updated_at.isoformat(),
        }
